File: programmers/level_1/64061.cpp
#include <string>
#include <vector>
#include <stack>

using namespace std;

int solution(vector<vector<int>> board, vector<int> moves) {
    int answer = 0;

    stack<int> st[30];
    for (int i = board[0].size() - 1; i >= 0; i--) {
        for (int j = 0; j < board[0].size(); j++) {
            if (board[i][j] != 0) {
                st[j].push(board[i][j]);
            }
        }
    }

    stack<int> s;
    for (int i = 0; i < moves.size(); i++) {
        int m = moves[i] - 1;
        if (!st[m].empty()) {
            int n = st[m].top();
            st[m].pop();

            if (!s.empty()) {
                int nn = s.top();
                if (nn == n) {
                    answer += 2;
                    s.pop();
                } else {
                    s.push(n);
                }
            } else {
                s.push(n);
            }
        }
    }

    return answer;
}

File: kotlin_ps/src/main/kotlin/programmers_64061/64061.kt
package programmers_64061

class Solution {
    fun solution(board: Array<IntArray>, moves: IntArray): Int {
        var answer = 0

        val topRows = IntArray(board.size)
        for (c in board.indices) {
            for (r in board.indices) {
                if (board[r][c] != 0) {
                    topRows[c] = r
                    break
                }
            }
        }

        val list = mutableListOf<Int>()
        for (move in moves) {
            val c = move - 1
            if (topRows[c] != board.size) {
                list.add(board[topRows[c]][c])
                if (list.size >= 2 && list[list.size - 1] == list[list.size - 2]) {
                    repeat(2) {
                        list.removeLast()
                        answer++
                    }
                }
                board[topRows[c]][c] = 0
                topRows[c]++
            }
        }

        return answer
    }
}